storage: let preemptive snapshots go through Raft

tbg opened the pull request:
Change the way in which preemptive snapshots are applied to use
a temporary Raft group. This lets Raft perform the verification
of the snapshot against its on-disk state and allows us to remove
some of the ad-hoc checks we have added recently.

Some fallout from this was making Raft config creation reusable,
trimming down `(*Replica).applySnapshot` and minor improvements
on the sending side.

See #6144.

cc @tamird @bdarnell @petermattis @arjunravinarayan

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7833)

<!-- Reviewable:end -->

bdarnell commented:
Reviewed 1 of 1 files at r1, 1 of 2 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 576 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMenY7of_j3Stl2rriM:-KMenY7of_j3Stl2rriN:-1119884678) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_raftstorage.go#L576)):*
> ```Go
> 		// to apply, so that should never happen, but we check anyway.
> 		// TODO(tschottdorf): is this correct or did we concluce that Raft's
> 		// sanity checks assumed we weren't touching the acked log entries?
> ```

I believe raft's sanity checks here are sufficient and we don't need to do our own checking here now that we're passing through raft's validation.

---

*[storage/replica_raftstorage.go, line 590 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeU3eND_Ltf7hA-8JU:-KMeodXTjtyX1kpfaD3E:292072078) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_raftstorage.go#L590)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is that surprising? seems expected - an old snapshot is still useful if the remainder of the work can be done by replaying the log.
</details>
It's surprising and subtle because there are two terms involved: `raftpb.Message.Term`, which is the `HardState.Term` of the node that sent the snapshot, and `raftpb.Message.Snapshot.Metadata.Term`, which is the term at which the last entry in the snapshot was proposed. 

It's fine for `Metadata.Term` to be older than the current term of the node sending it (although it is unusual for the two to differ for long). The checks implemented in this method are more conservative than those used in raft itself (which are somewhat subtle), so I think it's best to remove them and rely on raft's implementation.

One problem, though, is that we need to set `Message.Term` when sending the preemptive snapshot. Raft skips the term check for any message with Term 0 (on the assumption that it's a local message for which terms are irrelevant).

---

*[storage/store.go, line 2228 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeOdYX_oICmwV1_-K-:-KMeu3AX8OT6t7gVyouf:358937673) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f341c867e76df0fb7809d84ff70e926f7169300/storage/store.go#L2228)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this was confusing. it's not a replica ID, it's a raft ID.
</details>
Raft's node IDs are cockroach replica IDs.

---

*[storage/store.go, line 2226 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMerrlXk7b-OFTo47pV:-KMerrlYOe9FOrC8MszU:-1954209761) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2226)):*
> ```Go
> 			//
> 			// TODO(tschottdorf): This should be synchronized better with the
> 			// rest of the game. We're not under any continuous lock here. It's
> ```

This is under processRaftMu, which provides pretty strong synchronization. This is one of the areas we'll need to be careful about as we increase concurrency and break up processRaftMu, but for now it's safe.

---

*[storage/store.go, line 2230 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMesw9UMbeNrUn0QHqS:-KMesw9UMbeNrUn0QHqT:1170548742) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2230)):*
> ```Go
> 			// the preemptive snapshot and that we might clash into someone
> 			// else's keyspace.
> 			if !r.store.canApplySnapshot(req.RangeID, req.Message.Snapshot) {
> ```

We've already called canApplySnapshot above; no need to do it again.

---

*[storage/store.go, line 2259 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMetkRbVOH2LOBO17SZ:-KMetkRbVOH2LOBO17S_:-2103847786) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2259)):*
> ```Go
> 				if r.mu.internalRaftGroup != nil {
> 					// This can happen if a previous ChangeReplicas failed.
> 					return errors.New("cannot apply preemptive snapshot, Raft group already active")
> ```

Why can't we just apply the preemptive snapshot on the normal code path for initialized replicas? (i.e. instead of going down this preemptive snapshot path for ReplicaID==0, only do it when internalRaftGroup is nil)

---

*[storage/store.go, line 2262 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeu6v0VqD6D4B7M1lb:-KMeu6v0VqD6D4B7M1lc:1328759275) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2262)):*
> ```Go
> 				}
> 				const preemptiveSnapReplicaID = math.MaxUint64
> 				if r.mu.internalRaftGroup, err = raft.NewRawNode(&raft.Config{
> ```

I don't think this needs to be assigned to `r.mu.internalRaftGroup`. We can just create a local RawNode and call Step() on it.

---

*[storage/store.go, line 2266 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeuLkC2uQ4E5HxDjQH:-KMeuLkDUP6SHcNiQ7Ee:-2005646374) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2266)):*
> ```Go
> 					// TODO(tschottdorf): I don't think we have to be truthful
> 					// about this, but if we did have on-disk state, this Raft
> 					// instance might try to give us a large slab of entries
> ```

I don't think we need to set Applied here; raft will increase Applied for us when it applies the snapshot.

---

*[storage/store.go, line 2274 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeucMdi4vBEqWdrMdz:-KMeucMdi4vBEqWdrMe-:-528136234) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2274)):*
> ```Go
> 					// TODO(tschottdorf): copied verbatim from `withRaftGroup`,
> 					// should centralize this.
> 					ElectionTick:    r.store.ctx.RaftElectionTimeoutTicks,
> ```

Some of these can also be dummy values since we'll never call Tick() or other methods on this raft group. 

---

*[storage/store.go, line 2290 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeunvEdujUpPdRlDjX:-KMeunvEdujUpPdRlDjY:1995433967) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2290)):*
> ```Go
> 				// In the normal case, the group should have something to say.
> 				// If it doesn't, our snapshot was probably stale.
> 				if !r.mu.internalRaftGroup.HasReady() {
> ```

In some cases if the snapshot is discarded by raft we would get a Ready, but its Snapshot field would be empty.

---

*[storage/store.go, line 2299 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMev94zggAFRId_bn6q:-KMev94zggAFRId_bn6r:-279196054) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2299)):*
> ```Go
> 			// Now handle the reaction of Raft to receiving the snapshot
> 			// (usually applying the snapshot, along with a HardState update).
> 			if err := r.handleRaftReady(); err != nil {
> ```

We don't want to handle the whole Ready struct - in particular if it contains any Messages, we don't want to send them (they'll contain our bogus ReplicaID). We also don't want to accept a new SoftState because it assumes that the sender of a MsgSnap is a leader. We only want to process the snapshot and HardState.

---

*[storage/store.go, line 2303 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMevcuz7BgF9-39is-Q:-KMevcuz7BgF9-39is-R:-995498249) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2303)):*
> ```Go
> 			}
> 
> 			// TODO(tschottdorf): at this point, the Replica should simply
> ```

I don't think we want the Replica to disappear - we've had problems before when there was data on disk with no corresponding in-memory Replica object.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2303 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMevcuz7BgF9-39is-Q:-KMeyZhzxMrj7zQS00Xu:1998372956) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2303)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't think we want the Replica to disappear - we've had problems before when there was data on disk with no corresponding in-memory Replica object.
</details>
Context: https://github.com/cockroachdb/cockroach/pull/4204

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 590 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeU3eND_Ltf7hA-8JU:-KMgloUt_loerIsvz34m:1750343142) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_raftstorage.go#L590)):*
> One problem, though, is that we need to set `Message.Term` when sending the preemptive snapshot. Raft skips the term check for any message with Term 0 (on the assumption that it's a local message for which terms are irrelevant).

Is this documented? I think we should seriously consider adding preemptive snapshot support upstream, or we'll never stop depending on knowledge of upstream's implementation.

---

*[storage/store.go, line 2228 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeOdYX_oICmwV1_-K-:-KMgmNvWMHcoIJNZnfvq:963295460) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f341c867e76df0fb7809d84ff70e926f7169300/storage/store.go#L2228)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Raft's node IDs are cockroach replica IDs.
</details>
Right, but this particular value is not a cockroach replica ID; it's a completely bogus value which is _only_ a raft ID.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 2929 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeNxN42JeUfIjwzUhk:-KMhhccTSFtBvcPjNlnA:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/984cbb79703625f41574a964a78e78d93591ee68/storage/replica_command.go#L2929)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wrong place, obviously
</details>
Done

---

*[storage/replica_command.go, line 2941 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeOZerd5io5ovC7Qcn:-KMhhqjFHTrvrrbrOv05:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f341c867e76df0fb7809d84ff70e926f7169300/storage/replica_command.go#L2941)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> needs to include the replica ID itself
</details>
Done.

---

*[storage/replica_raftstorage.go, line 576 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMenY7of_j3Stl2rriM:-KMhnYVbpFTUAn8YLlyK:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_raftstorage.go#L576)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I believe raft's sanity checks here are sufficient and we don't need to do our own checking here now that we're passing through raft's validation.
</details>
Done.

---

*[storage/replica_raftstorage.go, line 590 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeU3eND_Ltf7hA-8JU:-KMhRyqg62UMtiJdCVM8:-400905814) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_raftstorage.go#L590)):*
> One problem, though, is that we need to set Message.Term when sending the preemptive snapshot.

Done.

> (on the assumption that it's a local message for which terms are irrelevant).

Shouldn't be hard to add an assertion against this somewhere down low enough?

---

*[storage/store.go, line 2228 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeOdYX_oICmwV1_-K-:-KMhVSLdZGFewzmhYcOx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f341c867e76df0fb7809d84ff70e926f7169300/storage/store.go#L2228)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Right, but this particular value is not a cockroach replica ID; it's a completely bogus value which is _only_ a raft ID.
</details>
Done.

---

*[storage/store.go, line 2226 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMerrlXk7b-OFTo47pV:-KMho6UNyHVyevsLzOP2:-1394792372) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2226)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is under processRaftMu, which provides pretty strong synchronization. This is one of the areas we'll need to be careful about as we increase concurrency and break up processRaftMu, but for now it's safe.
</details>
Removed parts of this comment. I'm still not comfortable with the locking here, but I suppose I need to look at it more.

---

*[storage/store.go, line 2227 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeUNokvgd5gI-Lkd4J:-KMhVFih78ET4LkD9RZF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2227)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> proceed?
</details>
Done.

---

*[storage/store.go, line 2230 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMesw9UMbeNrUn0QHqS:-KMhVG6uDDO4f-0I31W-:-273190089) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2230)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We've already called canApplySnapshot above; no need to do it again.
</details>
Oh, overlooked that. Done.

---

*[storage/store.go, line 2254 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeVE1O79S9H8Ig1Yrx:-KMhoC7Ri0nwJzvCyMm1:-1574352465) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2254)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> looks like this can avoid checking the error and instead always:
> ```
> 
> 			// TODO(tschottdorf): at this point, the Replica should simply
> 			// disappear. Since this is a WIP, we just nil out the group.
> 			r.mu.Lock()
> 			r.mu.internalRaftGroup = nil // force recreation on next op
> 			r.mu.Unlock()
> ```
</details>
This code is gone.

---

*[storage/store.go, line 2258 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeUgXpUw25SKNnWnaS:-KMhoOKs12b9HDB4EptS:58796457) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2258)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> more words?
</details>
Nope, less words. Code is gone.

---

*[storage/store.go, line 2259 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMetkRbVOH2LOBO17SZ:-KMhoUYQpC_RebiRJGyo:1101578519) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2259)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why can't we just apply the preemptive snapshot on the normal code path for initialized replicas? (i.e. instead of going down this preemptive snapshot path for ReplicaID==0, only do it when internalRaftGroup is nil)
</details>
I just didn't want to deal with this situation in the first version. Current version handles this (because it went back to never creating internalRaftGroup for a ReplicaIDReceiver).

---

*[storage/store.go, line 2262 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeu6v0VqD6D4B7M1lb:-KMhVTMxRCtnp5xSWkzD:-1712869437) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2262)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't think this needs to be assigned to `r.mu.internalRaftGroup`. We can just create a local RawNode and call Step() on it.
</details>
This was necessary because I was calling `handleRaftReady` which uses the raft group on `*Replica` in the first iteration. Current one doesn't assign any more and calls `applySnapshot` directly using the `raft.Ready`.

---

*[storage/store.go, line 2266 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeuLkC2uQ4E5HxDjQH:-KMhVWvDahiTBsi1oJvI:1566769464) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2266)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't think we need to set Applied here; raft will increase Applied for us when it applies the snapshot.
</details>
See the comment - I'm worried about the case in which we have some older on-disk state (say applied index 10k) and then we make a Raft group with 0 here. Wouldn't Raft spit a lot of entries back at us which it wants to apply? That seems good to avoid.

---

*[storage/store.go, line 2274 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeucMdi4vBEqWdrMdz:-KMhowbROon51F0sW-Nq:-19381231) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2274)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Some of these can also be dummy values since we'll never call Tick() or other methods on this raft group. 
</details>
I refactored out the config creation, mostly to avoid having to deal with these settings.

---

*[storage/store.go, line 2290 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeunvEdujUpPdRlDjX:-KMhp0FttlwzXS3XKNhD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2290)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> In some cases if the snapshot is discarded by raft we would get a Ready, but its Snapshot field would be empty.
</details>
Done.

---

*[storage/store.go, line 2299 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMev94zggAFRId_bn6q:-KMhp1a6LvasGXgYelg-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2299)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We don't want to handle the whole Ready struct - in particular if it contains any Messages, we don't want to send them (they'll contain our bogus ReplicaID). We also don't want to accept a new SoftState because it assumes that the sender of a MsgSnap is a leader. We only want to process the snapshot and HardState.
</details>
Done.

---

*[storage/store.go, line 2303 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMevcuz7BgF9-39is-Q:-KMh__KUC5mbHQI4JzhZ:616484440) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2303)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Context: https://github.com/cockroachdb/cockroach/pull/4204
</details>
Comments updated to reflect that brainfart.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 272 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMismCmHGYKo9nnsTAb:-KMismCnYUq4CCkXy7Rz:35055655) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/replica.go#L272)):*
> ```Go
> 	if r.mu.internalRaftGroup == nil {
> 		raftGroup, err := raft.NewRawNode(newRaftConfig(
> 			raft.Storage(r),
> ```

don't think you need this raft.Storage()

---

*[storage/store.go, line 2304 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMiu2wOWAY4ICk52WD4:-KMiu2wOWAY4ICk52WD5:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2304)):*
> ```Go
> 			raftGroup, err := raft.NewRawNode(
> 				newRaftConfig(
> 					raft.Storage(r),
> ```

ditto

---

*[storage/store.go, line 2330 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMiuXZiTLQvHFVRSUZP:-KMiuXZiTLQvHFVRSUZQ:1531025487) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2330)):*
> ```Go
> 				return err
> 			}
> 			// TODO(tschottdorf): at this point, the Replica has data but
> ```

what's the TODO here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[storage/client_raft_test.go, line 1361 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMf1Io5x4W0nbbK7lfE:-KMf1Io5x4W0nbbK7lfF:-295737274) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3988422b2616a5270fd49af76285b1b8e748e72/storage/client_raft_test.go#L1361)):*
> ```Go
> 	// TODO(peter): We're sometimes generating normal snapshots immediately after
> 	// the preemptive ones. Need to figure out why and fix.
> 	if false {
> ```

This test was disabled because it was flaky. Sometimes non-preemptive snapshots occur and I'm pretty sure it is due to the async sending of preemptive snapshots. 

---

*[storage/replica_command.go, line 2943 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMf1kjV_dGC04UD5gp1:-KMf1kjV_dGC04UD5gp2:1769717425) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_command.go#L2943)):*
> ```Go
> 			return errors.Errorf("must not specify a ReplicaID for new Replica")
> 		}
> 		r.raftSender.SendAsync(&RaftMessageRequest{
> ```

See my comment on the preemptive snapshot test. I think this should really be a synchronous RPC. Even better, we should make this a streaming RPC and send the snapshot in bits, waiting for the remote to write it to disk. Both these items can wait for subsequent PRs.

---

*[storage/replica_raftstorage.go, line 631 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjIjnps5F0jOYhQIc-:-KMjIjnps5F0jOYhQIc0:-1398835697) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/replica_raftstorage.go#L631)):*
> ```Go
> 
> 	batch.Defer(func() {
> 		// As the last deferred action after committing the batch, update other
> ```

If you want to be guaranteed this is the last deferred action, you should put this defer up right after the creation of the batch. Otherwise it is easy for someone else to add a `batch.Defer` before this one. 

Is there a reason you prefer using a `batch.Defer` for this? The reverse order execution of defers, while symmetric with Go's builtin `defer` is unintuitive to the casual reader. 

---

*[storage/store.go, line 2327 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjP00ZMUR9CftaOJxx:-KMjP00ZMUR9CftaOJxy:-762447284) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2327)):*
> ```Go
> 			}
> 			// Apply the snapshot, as Raft told us to.
> 			if err := r.applySnapshot(ready.Snapshot, ready.HardState); err != nil {
> ```

Just noticed that applying a snapshot here will prevent other raft processing on the store. When we eventually make normal raft processing concurrent in `Store.processRaft`, this will be a bottleneck. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 590 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeU3eND_Ltf7hA-8JU:-KMjV1m0TI91ev0i4fwD:-1387834826) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_raftstorage.go#L590)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> > One problem, though, is that we need to set Message.Term when sending the preemptive snapshot.
> 
> Done.
> 
> > (on the assumption that it's a local message for which terms are irrelevant).
> 
> Shouldn't be hard to add an assertion against this somewhere down low enough?
</details>
Nothing about raft.Message is currently documented; any time we construct one (here and for preemptive snapshots) we're relying on implementation details. Once we have a plan for streaming snapshots we should look at changes to the upstream interfaces to support both streaming and preemptive snapshots. 

I think we could also tighten up some assertions internal to raft to guard against improperly hand-crafted Messages. Currently it makes a lot of assumptions, like assuming a message without a term is a message that doesn't need one (instead of a mistake). 

---

*[storage/store.go, line 2266 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeuLkC2uQ4E5HxDjQH:-KMj_0ma5Kip4cZzUCNW:131584463) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2266)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> See the comment - I'm worried about the case in which we have some older on-disk state (say applied index 10k) and then we make a Raft group with 0 here. Wouldn't Raft spit a lot of entries back at us which it wants to apply? That seems good to avoid.
</details>
No, because the snapshot updates the applied index before raft decides whether it has new entries for us to apply. After a snapshot `Applied == Commit == LastIndex` and so there are no new entries for it to give us.

---

*[storage/store.go, line 2265 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZQPrCEc05N5VqX6s:-KMjZQPrCEc05N5VqX6t:-1923324511) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2265)):*
> ```Go
> 			// at term zero for internal messages). The sending side uses the
> 			// term from the snapshot itself, so we check for equality here.
> 			if a, e := req.Message.Term, req.Message.Snapshot.Metadata.Term; a != e {
> ```

It's not required that `Message.Term == Snapshot.Metadata.Term`. I think it is required that `Message.Term >= Snapshot.Metadata.Term`, but why do this check yourself instead of just letting raft do it?

---

*[storage/store.go, line 2297 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZmLC7zZy8BTk_sFz:-KMjZmLD-HDOoS6z-APm:348561619) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2297)):*
> ```Go
> 			r.mu.Unlock()
> 			if groupExists {
> 				log.Fatalf(
> ```

If you want to enforce this (and I still think it's better to go through the regular code path when a raft group exists), it shouldn't be a `Fatalf`. This is possible in the real world if a replica is removed and re-added before it can be GC'd.

---

*[storage/store.go, line 2327 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjP00ZMUR9CftaOJxx:-KMj_bdRTY90UA8Xspj0:1860396985) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2327)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Just noticed that applying a snapshot here will prevent other raft processing on the store. When we eventually make normal raft processing concurrent in `Store.processRaft`, this will be a bottleneck. 
</details>
The solution to #7830 should let us fix this. Once we can reserve our spot in `replicasByKey` before applying the snapshot, we no longer need to hold `processRaftMu` throughout the snapshot application process, and we can use a Replica-level lock instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I started this branch as an experiment (to verify our assumption that we could avoid out-of-Raft snapshots), and that seems to have worked out well. Personally, I don't feel that we should re-enable preemptive snapshots due to the related-but-out-of-scope concurrency issues (#7830), so my suggestion is to land this but take out the temporary commit which re-enabled preemptive snapshots until we have those issues sorted out. See also https://github.com/cockroachdb/cockroach/issues/7875 about removing manual HardState updates (of which I removed one in this PR).

---

Review status: 1 of 7 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[storage/client_raft_test.go, line 1361 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMf1Io5x4W0nbbK7lfE:-KMxARXphS53ZNEYfHDx:-1616651335) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3988422b2616a5270fd49af76285b1b8e748e72/storage/client_raft_test.go#L1361)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This test was disabled because it was flaky. Sometimes non-preemptive snapshots occur and I'm pretty sure it is due to the async sending of preemptive snapshots. 
</details>
This change is in a commit I have titled `[temp]` for that reason. We'll have to discuss whether we actually want to re-enable preemptive snapshots at this point, but at least during development I wanted to run this test.

---

*[storage/replica.go, line 272 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMismCmHGYKo9nnsTAb:-KMxAeb6vw00TcVH0iXp:-1469505312) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/replica.go#L272)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> don't think you need this raft.Storage()
</details>
I added it intentionally since it clarifies what we're passing the Replica as.

---

*[storage/replica_command.go, line 2943 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMf1kjV_dGC04UD5gp1:-KMxAntCDUkr74zolTMO:-1955495354) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/replica_command.go#L2943)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> See my comment on the preemptive snapshot test. I think this should really be a synchronous RPC. Even better, we should make this a streaming RPC and send the snapshot in bits, waiting for the remote to write it to disk. Both these items can wait for subsequent PRs.
</details>
Yep, I only enabled this for development. I'll see whether it's easy to turn this into a sync RPC (thus reenabling the test and, if we decide we want to, preemptive snapshots in general) when the rest of the change is approved.

---

*[storage/replica_raftstorage.go, line 631 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjIjnps5F0jOYhQIc-:-KMxB7cqAUlooMdT-jOH:2040689167) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/replica_raftstorage.go#L631)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> If you want to be guaranteed this is the last deferred action, you should put this defer up right after the creation of the batch. Otherwise it is easy for someone else to add a `batch.Defer` before this one. 
> 
> Is there a reason you prefer using a `batch.Defer` for this? The reverse order execution of defers, while symmetric with Go's builtin `defer` is unintuitive to the casual reader. 
</details>
I have a recent tendency to group side effects into `Defer`s since I'm planning to remove `batch.Defer` (which means that I'm going to see all the side effects using one in the process).

---

*[storage/store.go, line 2266 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeuLkC2uQ4E5HxDjQH:-KMxC56OIZy1IikSKt2x:-439962368) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2266)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> No, because the snapshot updates the applied index before raft decides whether it has new entries for us to apply. After a snapshot `Applied == Commit == LastIndex` and so there are no new entries for it to give us.
</details>
That's after a snapshot which is applied. The snapshot could be discarded.
Well, I suppose it's rare enough to not worry about, even if it does happen every once in a blue moon.

---

*[storage/store.go, line 2265 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZQPrCEc05N5VqX6s:-KMxBMxjwPzWggRrUy-2:298646223) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2265)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It's not required that `Message.Term == Snapshot.Metadata.Term`. I think it is required that `Message.Term >= Snapshot.Metadata.Term`, but why do this check yourself instead of just letting raft do it?
</details>
This is really just guarding against messages at Term zero. Since we handcraft this message at the sender using the snapshot's term, it seems silly not to check for this exact condition here as well. I thought the comment above explained that well, did you just miss it?

---

*[storage/store.go, line 2297 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZmLC7zZy8BTk_sFz:-KMxCf9IfBgwD500lpma:-1077701088) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2297)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> If you want to enforce this (and I still think it's better to go through the regular code path when a raft group exists), it shouldn't be a `Fatalf`. This is possible in the real world if a replica is removed and re-added before it can be GC'd.
</details>
You're right about the `Fatalf`, of course.
What do you mean by "going through the regular code path"? You want `raftGroup := r.mu.internalRaftGroup`? That makes the locking more complicated since now I have to (or at least should) lock `*Replica.mu` in one of the branches (but I can't hold the lock throughout since `handleRaftReady` needs it).
Or you want to leave the `ReplicaID==0` code path? That's also tricky since we don't want to apply all the parts of the `Ready`.
The case which you describe above is rare enough to discard the occasional snapshot, so I'm inclined to keep rejecting the snapshot in that case just to keep things stupid and simple. But I'm curious what you were asking for.

---

*[storage/store.go, line 2330 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMiuXZiTLQvHFVRSUZP:-KMxCTbaTGdhjhOAERpn:2137327179) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2330)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's the TODO here?
</details>
Turned this into a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-:-KMxFrH1iI5tehOW5EZJ:978988178)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Why is #7830 a reason to block preemptive snapshots? I don't see how this makes us any more vulnerable to overlapping ranges than we were before. 

---

Reviewed 1 of 7 files at r9, 6 of 6 files at r10.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2259 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMetkRbVOH2LOBO17SZ:-KMzinueYZ8nfF0d2-vJ:-1112936350) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2259)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I just didn't want to deal with this situation in the first version. Current version handles this (because it went back to never creating internalRaftGroup for a ReplicaIDReceiver).
</details>
This is better, but you're still dealing with the consequences of a raft group that already exists (the `groupExists` check above). It seems simpler to me to handle that by passing the snapshot to Step 

---

*[storage/store.go, line 2266 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeuLkC2uQ4E5HxDjQH:-KMzmb3vYRZaH7PWV3cw:2028818284) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2266)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> That's after a snapshot which is applied. The snapshot could be discarded.
> Well, I suppose it's rare enough to not worry about, even if it does happen every once in a blue moon.
</details>
Ah, that's subtle. Yes, if the snapshot is discarded and we have `Applied: 0`, then raft will attempt to replay the whole log into `Ready.CommittedEntries`. It will fail (and panic), because the log doesn't start at zero. (our `Replica.Entries` method is kind of stupid and will load all the log entries into memory before realizing this and returning an error). 

I think this TODO has gotten lost. 

---

*[storage/store.go, line 2265 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZQPrCEc05N5VqX6s:-KMzhxA7lTdfSofdj6K6:645004994) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2265)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This is really just guarding against messages at Term zero. Since we handcraft this message at the sender using the snapshot's term, it seems silly not to check for this exact condition here as well. I thought the comment above explained that well, did you just miss it?
</details>
Yeah, I missed that part of the comment. I still think this is unnecessarily strict and would prefer to just check that `req.Message.Term != 0` and let raft do the rest.

---

*[storage/store.go, line 2297 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZmLC7zZy8BTk_sFz:-KMziEwGt7kOw4dZvfLh:-2025823586) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2297)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You're right about the `Fatalf`, of course.
> What do you mean by "going through the regular code path"? You want `raftGroup := r.mu.internalRaftGroup`? That makes the locking more complicated since now I have to (or at least should) lock `*Replica.mu` in one of the branches (but I can't hold the lock throughout since `handleRaftReady` needs it).
> Or you want to leave the `ReplicaID==0` code path? That's also tricky since we don't want to apply all the parts of the `Ready`.
> The case which you describe above is rare enough to discard the occasional snapshot, so I'm inclined to keep rejecting the snapshot in that case just to keep things stupid and simple. But I'm curious what you were asking for.
</details>
I mean that we should only enter the preemptive snapshot branch if `req.ToReplica.ReplicaID == 0` *and* `r.mu.internalRaftGroup == nil`. If there is a non-nil raft group already, we should go on to call `raftGroup.Step` instead of following the path for preemptive snapshots.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-:-KMz_ndYvA_k60xj-P7T:-1917427895)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r9, 6 of 6 files at r10.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2259 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMetkRbVOH2LOBO17SZ:-KN1rct2L6l_H-JwjGTS:-560585436) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2259)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is better, but you're still dealing with the consequences of a raft group that already exists (the `groupExists` check above). It seems simpler to me to handle that by passing the snapshot to Step 
</details>
Hmm, didn't think you were suggesting feeding the preemptive snapshot message to `internalRaftGroup.Step`. That would make it pop out of `handleRaftMessage` and we talked earlier about not wanting to handleRaftReady to process that message (and it's definitely not simpler to have two possible full-fledged code paths that a preemptive snapshot could take). Perhaps it's just FUD, but feeding a hand-crafted message into code paths that have all the right to assume that the origin of the message is a real Raft peer doesn't sit well with me.

I'd prefer to let the preemptive snapshot die here if it comes in at an unexpected time, but I could be convinced if we understand the situation really, really well and it seems to have a worthwhile impact doing it.

---

*[storage/store.go, line 2266 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMeuLkC2uQ4E5HxDjQH:-KN1wq_E5A4sXhZaEolL:676965300) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2266)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Ah, that's subtle. Yes, if the snapshot is discarded and we have `Applied: 0`, then raft will attempt to replay the whole log into `Ready.CommittedEntries`. It will fail (and panic), because the log doesn't start at zero. (our `Replica.Entries` method is kind of stupid and will load all the log entries into memory before realizing this and returning an error). 
> 
> I think this TODO has gotten lost. 
</details>
PTAL

---

*[storage/store.go, line 2265 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZQPrCEc05N5VqX6s:-KN1rNZx1uKoozaGHlsY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2265)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, I missed that part of the comment. I still think this is unnecessarily strict and would prefer to just check that `req.Message.Term != 0` and let raft do the rest.
</details>
Done.

---

*[storage/store.go, line 2297 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjZmLC7zZy8BTk_sFz:-KN1xGjed_5ivmPgH1aZ:400037716) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2297)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I mean that we should only enter the preemptive snapshot branch if `req.ToReplica.ReplicaID == 0` *and* `r.mu.internalRaftGroup == nil`. If there is a non-nil raft group already, we should go on to call `raftGroup.Step` instead of following the path for preemptive snapshots.
</details>
See below, I'll need a bit of convincing that this is a) safe b) worth it.

---

*[storage/store.go, line 2327 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMjP00ZMUR9CftaOJxx:-KN1xSD00989ab3WuDKP:-1608743340) ([raw file](https://github.com/cockroachdb/cockroach/blob/420a7bb6f4abc678690f235d52246aa30ddda692/storage/store.go#L2327)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The solution to #7830 should let us fix this. Once we can reserve our spot in `replicasByKey` before applying the snapshot, we no longer need to hold `processRaftMu` throughout the snapshot application process, and we can use a Replica-level lock instead.
</details>
Pinged #7830 about this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Couldn't the fact that a preemptive snapshot can arrive roughly along with a corresponding Raft snapshot cause issues? Perhaps not (due to them having the same range), but still. If you're more comfortable re-enabling preemptive snapshots, I'm going to defer to that judgement (in which case I should begin fixing up the corresponding tests).

---

Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-:-KN21pRispc4te9Bjl2y:-1551006913)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I seem to have introduced a test failure:

`make stress PKG=./storage TESTS=TestGossipFirstRange TESTFLAGS='-test.timeout 10s'`

fails after <1k iters. Just a reminder to myself to sort out, may not be a big deal.

petermattis commented:
@tschottdorf I just added `TestGossipFirstRange`. I'm curious how/where it is failing with your PR?

tschottdorf commented:
https://gist.github.com/tschottdorf/aca3521a748ba257171fc528a21aa245

petermattis commented:
Hmm, the test is failing at `gossip_test.go:101` and then hanging when calling the defered `tc.Stopper().Stop()`. I don't have an explanation for either problem. I'll try reproducing locally without your PR.

tschottdorf commented:
It doesn't repro without the PR (at least not within 20k iters on GCE when it failed in <1k with the PR).

petermattis commented:
@tschottdorf You're correct that it doesn't repro without this branch. Do you need help investigating the failure?

tschottdorf commented:
I'm sidetracked looking at Jepsen, so if you have spare capacity I'd
appreciate it.

On Wed, Jul 20, 2016 at 9:47 AM Peter Mattis <notifications@github.com>
wrote:

> @tschottdorf <https://github.com/tschottdorf> You're correct that it
> doesn't repro without this branch. Do you need help investigating the
> failure?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7833#issuecomment-233953825>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135E6aBOm277JvUCilX8VdIhKr4wUEks5qXib_gaJpZM4JMocB>
> .
>
-- 

-- Tobias

petermattis commented:
Ok. I can reproduce the problem locally using your branch.

petermattis commented:
The `TestGossipFirstRange` failure looks like a race in the test. I'm not seeing why it doesn't occur outside of this PR, since it should. The specific problem is that the `changeReplicasTrigger` for the replica addition can run after the lease is transferred to node 2 even though it logically occurs before it in the test. When this happens, we gossip the first range once more than the test expects. I don't think there is any easy way to get rid of this extra unnecessary gossip. Simplest approach is to fix the test.

tschottdorf commented:
While I wait for @bdarnell's feedback, I'll be fixing the preemptive snapshot tests (tomorrow) since the general consensus seems to be that we do want to enable them again.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for the `TestStoreRangeRebalance` change. We're probably not going to be able to make this non-flaky until we introduce some sort of acknowledgement for snapshots into the protocol, so I'd say we merge this with that test disabled and fix it in a separate PR.

---

Reviewed 6 of 6 files at r12.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2259 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-KMetkRbVOH2LOBO17SZ:-KNPbEKzQvyKOUbg2s9o:-286932417) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc0f6bfe61e90baa12e5f38a115369f1323745d/storage/store.go#L2259)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Hmm, didn't think you were suggesting feeding the preemptive snapshot message to `internalRaftGroup.Step`. That would make it pop out of `handleRaftMessage` and we talked earlier about not wanting to handleRaftReady to process that message (and it's definitely not simpler to have two possible full-fledged code paths that a preemptive snapshot could take). Perhaps it's just FUD, but feeding a hand-crafted message into code paths that have all the right to assume that the origin of the message is a real Raft peer doesn't sit well with me.
> 
> I'd prefer to let the preemptive snapshot die here if it comes in at an unexpected time, but I could be convinced if we understand the situation really, really well and it seems to have a worthwhile impact doing it.
</details>
My thinking is that we're already committed to hand-crafting `raftpb.Messages` that raft can handle correctly (for coalesced heartbeats); this part worries me less than constructing the raft group with a bogus ID. But you're right that this is unlikely to do much good so it's safer to drop it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7833#-:-KNPeW7zxRE-_myRnWOm:-575659924)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: let preemptive snapshots go through Raft

Change the way in which preemptive snapshots are applied to use
a temporary Raft group. This lets Raft perform the verification
of the snapshot against its on-disk state and allows us to remove
some of the ad-hoc checks we have added recently.

Some fallout from this was making Raft config creation reusable,
trimming down `(*Replica).applySnapshot` and minor improvements
on the sending side.

Re-enabled preemptive snapshots, though the corresponding test
is still flaky and remains disabled. It is already tracked
separately.

See #6144.

tbg committed:
de-flake TestTransferRaftLeadership

The test as written invited flakiness and seemed to have
successfully procured it with the preemptive snapshot changes
when combined with a lease acquisition in the replica change trigger.
